realm-transformer/src/main/groovy/io/realm/transformer/BytecodeModifier.groovy

```diff
@@ -147,4 +149,15 @@ class BytecodeModifier {
                     new CtClass[0], new CtClass[0], "{return true;}", clazz))
         }
     }
+
+    private static CtMethod createSetter(String methodName, CtField field) {
+        CtMethod setter = CtNewMethod.setter(methodName, field)
+        if ("io.realm.RealmInteger".equals(field.getType().getName())) {
+            setter.insertBefore(
+                    'if (!($1 instanceof io.realm.internal.datatypes.realminteger.UnmanagedRealmInteger))' +
+                    'throw new IllegalStateException("Cannot assign a managed RealmInteger to an unmanaged object");');
```

Comment:
I have been convinced on this.  As ugly as it seems, allowing assignment of a managed RealmInteger, to a field in an unmanaged RealmObject, is, as @zaki50 points out, more consistent with the behaviour of the rest of our API.